cheatsheets_draft/GraphQL_Cheat_Sheet.md

```diff
@@ -0,0 +1,257 @@
+- [DoS](https://owasp.org/www-community/attacks/Denial_of_Service) ([Denial of Service](https://www.cloudflare.com/learning/ddos/glossary/denial-of-service/))
+- [IDOR](https://portswigger.net/web-security/access-control/idor)
+- Broken authorization: either [improper](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa1-broken-object-level-authorization.md) or [excessive](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa3-excessive-data-exposure.md) access
+
+## Best Practices and Recommendations
+
+### Input Validation
+
+Adding strict input validation can help prevent against injection and DoS. The main design for GraphQL is that the user supplies one or more identifiers and the backend has a number of data fetchers making HTTP, DB, or other calls using the given identifiers. This means that user input will be included in HTTP requests, DB queries, or other requests/calls which provides opportunity for injection that could lead to various injection attacks or DoS.
+
+See the OWASP Cheat Sheets on [Input Validation](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html) and general [injection prevention](https://cheatsheetseries.owasp.org/cheatsheets/Injection_Prevention_Cheat_Sheet.html) for full details to best perform input validation and prevent injection.
+
+#### General Practices
+
+Validate all incoming data to only allow valid values (i.e. whitelist).
+
+- Use specific GraphQL [data types](https://graphql.org/learn/schema/#type-language) such as [scalars](https://graphql.org/learn/schema/#scalar-types) or [enums](https://graphql.org/learn/schema/#enumeration-types). Write custom GraphQL [validators](https://graphql.org/learn/validation/) for more complex validations. [Custom scalars](https://itnext.io/custom-scalars-in-graphql-9c26f43133f3) may also come in handy.
+- Define [schemas for mutations input](https://graphql.org/learn/schema/#input-types).
+- [Whitelist allowed characters](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#whitelisting-vs-blacklisting) - don't use a blacklist
+    - The stricter the whitelist the better. A lot of times a good starting point is only allowing alphanumeric, non-unicode characters because it will disallow many attacks.
+- To properly handle unicode input, use a [single internal character encoding](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#validating-free-form-unicode-text)
+- Gracefully [reject invalid input](https://cheatsheetseries.owasp.org/cheatsheets/Error_Handling_Cheat_Sheet.html), being careful not to reveal excessive information about how the API and its validation works.
+
+#### Injection Prevention
+
+When handling input meant to be passed to another interpreter (e.g. SQL/NoSQL/ORM, OS, LDAP, XML):
+
+- Always prefer safe tools with support for parameterized statements
```

Comment:
We won't be able to define what "safe tools" are, but "safe APIs" is used by other OWASP documents (e.g. [A1:2017 - Injection](https://owasp.org/www-project-top-ten/OWASP_Top_Ten_2017/Top_10-2017_A1-Injection))

```suggestion
- Always choose libraries/modules/packages offering safe APIs, such as parameterized statements.
```

Reply:
Yeah I wanted something more generic than "APIs" because I think that term is too heavily used to reference HTTP services and didn't want to be confusing. But your new suggestion works great! I'll add that in.